imblearn/over_sampling/_smote.py

```diff
@@ -1090,3 +1093,186 @@ def _generate_sample(self, X, nn_data, nn_num, row, col, step):
             sample[start_idx + col_sel] = 1
 
         return sparse.csr_matrix(sample) if sparse.issparse(X) else sample
+
+
+@Substitution(
+    sampling_strategy=BaseOverSampler._sampling_strategy_docstring,
+    random_state=_random_state_docstring)
+class KMeansSMOTE(BaseSMOTE):
+    """Apply a KMeans clustering before to over-sample using SMOTE.
+
+    This is an implementation of the algorithm described in [1]_.
+
+    Read more in the :ref:`User Guide <smote_adasyn>`.
+
+    Parameters
+    ----------
+    {sampling_strategy}
+
+    {random_state}
+
+    k_neighbors : int or object, optional (default=2)
+        If ``int``, number of nearest neighbours to used to construct synthetic
+        samples.  If object, an estimator that inherits from
+        :class:`sklearn.neighbors.base.KNeighborsMixin` that will be used to
+        find the k_neighbors.
+
+    n_jobs : int, optional (default=1)
+        The number of threads to open if possible.
+
+    kmeans_estimator : int or object, optional (default=KMeans())
+        A KMeans instance or the number of clusters to be used.
+
+    cluster_balance_threshold: str or float, optional (default="auto")
+        The threshold at which a cluster is called balanced and where samples
+        of the class selected for SMOTE will be oversampled. If "auto", this
+        will be determined by the ratio for each class, or it can be set
+        manually.
+
+    density_exponent: str or float, optional (default="auto")
+        This exponent is used to determine the density of a cluster. Leaving it
+        to 'auto' will use a feature-length based exponent.
+
+    ratio : str, dict, or callable
+        .. deprecated:: 0.4
+           Use the parameter ``sampling_strategy`` instead. It will be removed
+           in 0.6.
+
+    """
```

Comment:
At the end of the docstring we could add the reference of the paper and after that an interactive example. Check [here](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/master/imblearn/over_sampling/_smote.py#L281) for instance.

Reply:
I added an interactive example that specifically demonstrates the utility of the KMeansSmote class, 3 blobs, with the positive class in the middle and the negative classes on the outside and a single negative sample in the middle blob. The example shows that after resampling no new samples are added in the middle blob. Inspired by the following toy problem:
![image](https://user-images.githubusercontent.com/3799160/57215533-7311f600-6fed-11e9-97ed-e1339592b08c.png)